Add a quickstart rigid flap tutorial

DavidSCN opened the pull request:
Resolves #117 
Setup as described in #117 . We decided to remove the `nearest-projection` capability and use a RBF mappings instead.

![](https://user-images.githubusercontent.com/33414590/100848273-2d473900-3481-11eb-8a69-c53e4d28343d.gif)

DavidSCN commented on FSI/rigid_flap/OpenFOAM-RigidBody/Fluid/system/fvSolution:
```diff
@@ -0,0 +1,110 @@
+    pisoTol 1e-6;
+    consistent  true;
+
+    // OpenFOAM (.com), OpenFOAM 5 or older (.org)
```
I will check if we can remove this block completely since it is an unnecessary complication for newcomer.

uekerman reviewed (changes requested):
Nice :+1: 

* Let's also add a `export:vtk` to the Solid participant and explain in the docs how to visualize the point cloud.
* Could you please format the C++ file with the same clang-format options as used in preCICE?

The IQN convergence is still a bit bad. I'll try whether better settings are possible.

MakisH commented on FSI/rigid_flap/OpenFOAM-RigidBody/precice-config.xml:
```diff
@@ -0,0 +1,75 @@
+  </log>
+
+  <solver-interface dimensions="2">
+    <data:vector name="Force" />
+    <data:vector name="Displacement" />
```
I don't think we really have a consistent scheme right now. I took a note for our tutorials restructuring.

I also think this is not a trivial to argue decision: one could see it as "what is the quantity I couple" (thus singular), but when we speak we usually use plural, as it somehow sounds better. But I think that the singular (used here) is easier to keep consistent (works even on one-point interfaces) and more correct: when writing the mapping equations, one does not see a plural there. Moreover, we could use the plural to refer to the two sides of the interface (`Forces: Force_Fluid, Force_Solid`).

But overall, I would not address this right now, but only at the batch-editing phase.

DavidSCN commented on FSI/rigid_flap/OpenFOAM-RigidBody/Solid/rigid_body_solver.cpp:
```diff
@@ -0,0 +1,252 @@
+    // Define a boundary mesh
+    std::cout << "Rigid body: defining mesh: ";
+    std::cout << n_nodes << " nodes " << std::endl;
+    const double delta_y = height / (n_vertical_nodes - 1);
+    const double delta_x = length / (n_horizontal_nodes - 1);
```
They are here declared in a local scope, but I moved them up.

DavidSCN commented on FSI/rigid_flap/OpenFOAM-RigidBody/Solid/rigid_body_solver.cpp:
```diff
@@ -0,0 +1,252 @@
+    // Store rigid body angle theta^{n}
+    const double theta_old = theta;
+    // Update angle to theta^{n+1} according to forward Euler method (simplified moment
+    // computation, which does not depend on the updated configuration)
```
Done, please check again if this is helpful.

DavidSCN commented:
> Could you please format the C++ file with the same clang-format options as used in preCICE?

I used now the main repo .clang-format file and it should be formatted properly. I'm still a bit confused why the comments are not formatted (linebreaks), but I guess this is not configured in preCICE itself.

uekerman reviewed (approved):
Looks good. Let me only still check the IQN performance. I can merge then.

MakisH commented on FSI/rigid_flap/OpenFOAM-RigidBody/Fluid/system/decomposeParDict:
```diff
@@ -0,0 +1,25 @@
+}
+// * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * //
+
+    numberOfSubdomains 25;
```
Maybe use a more modest number here, so that it can easily be used on a laptop?

DavidSCN replied:
Oh I haven't noticed the file at all. I changed it to two.

MakisH commented on FSI/rigid_flap/OpenFOAM-RigidBody/Solid/rigid_body_solver.cpp:
```diff
@@ -0,0 +1,246 @@
+#include <iostream>
+#include <math.h>
```
Better use the C++ headers:
```suggestion
#include <cmath>
```

MakisH commented on FSI/rigid_flap/OpenFOAM-RigidBody/Solid/rigid_body_solver.cpp:
```diff
@@ -0,0 +1,246 @@
+
+  void reload_old_state(Vector &vertices,
+                        double &theta,
+                        double &theta_dot) const
```
Why declare this a const method? It should not work with const objects anyway.

DavidSCN replied:
Why not? I pulled the `markActionFulfilled` statement out in the main loop. So, it is actually a `const` method.

MakisH replied:
You are right, I quickly read it and understood `reload_old_state` as modifying the DataContainer (i.e. `old_vertices = vertices`). This is totally fine.

MakisH commented on FSI/rigid_flap/OpenFOAM-RigidBody/Solid/rigid_body_solver.cpp:
```diff
@@ -0,0 +1,246 @@
+        Vector &      vertices,
+        double &      theta,
+        double &      theta_dot,
+        const double  delta_t) const
```
same here for `const`

MakisH commented on FSI/rigid_flap/OpenFOAM-RigidBody/Solid/rigid_body_solver.cpp:
```diff
@@ -0,0 +1,246 @@
+  {
+    // Compute total moment M = x^{n} x f^{n+1}
+    double moment = 0;
+    for (uint i = 0; i < forces.size() / 2; ++i)
```
Just use `auto` here (where does actually `uint` come from?)

```suggestion
    for (auto i = 0; i < forces.size() / 2; ++i)
```

DavidSCN replied:
`unsigned int` ?

MakisH replied:
It looks like `uint` is compiler-specific: https://stackoverflow.com/questions/3552094/c-uint-unsigned-int-int

DavidSCN replied:
Oh, I was not aware, good to know

MakisH commented on FSI/rigid_flap/OpenFOAM-RigidBody/Solid/rigid_body_solver.cpp:
```diff
@@ -0,0 +1,246 @@
+#include <iostream>
```
We should also explicitly include `string` here, even if `iostream` includes some parts:

```suggestion
#include <string>
```

DavidSCN replied:
`iostream` is still required.

MakisH commented on FSI/rigid_flap/OpenFOAM-RigidBody/Solid/rigid_body_solver.cpp:
```diff
@@ -0,0 +1,246 @@
+                          vertex_ids.data());
+
+  // initialize the Solverinterface
+  double dt = precice.initialize();
```
In such cases where what is returned already has a type, you can use `auto` to avoid implicit conversions (although here this is clear).

MakisH reviewed (commented):
I quickly went through the code (did not try to run it) and I have a few minor suggestions. Feel free to integrate as many as you want. I can also try running it, if you want.

DavidSCN commented:
> Looks good. Let me only still check the IQN performance. I can merge then.

Where should the `README` go? I guess I should add it before merging this?

MakisH commented:
> > Looks good. Let me only still check the IQN performance. I can merge then.
> 
> Where should the `README` go? I guess I should add it before merging this?

Yes, please. Even if we keep the instructions in a separate place (already now or later), we need a minimal README here as well.

MakisH reviewed (commented):
- I tried running it (serial & parallel) and it works nicely. The Fluid results in ParaView look fine, the watchpoint plot was clear. The motion feels a bit too intense, but good that preCICE converges in such conditions.
- The code looks clean & nice :+1:

Two small todos:
- [ ] We still need a README, but you have this already in mind.
- [x] We still need to add the script `removeObsoleteFolders.sh`, similarly to other OpenFOAM tutorials.

MakisH commented:
> Looks good. Let me only still check the IQN performance. I can merge then.

Here are the convergence and iteration files for the parallel case:

[precice-Solid-convergence.log](https://github.com/precice/tutorials/files/5657953/precice-Solid-convergence.log)
[precice-Solid-iterations.log](https://github.com/precice/tutorials/files/5657954/precice-Solid-iterations.log)

It starts converging in 3-4 iterations, goes up to 7 iterations in the given time frame.

DavidSCN commented:
> The motion feels a bit too intense

It's indeed near an (even with implicit coupling) unstable configuration. We could reduce the motion by, e.g.,  modifying the Fluid density.

MakisH commented:
> We could reduce the motion by, e.g., modifying the Fluid density.

Or maybe reducing the (output) time step. We mainly want to change the visual effect here.

MakisH commented on FSI/rigid_flap/OpenFOAM-RigidBody/README.md:
```diff
@@ -0,0 +1,27 @@
+
+This tutorial deals with a fluid-structure-interaction problem. The fluid part of the simulation is computed using OpenFOAM and the rigid body motion is a rigid body model written in `c++` with only a single degree of freedom, namely the deflection angle of the flap in the channel. The rigid body is fixed in the origin at (0,0) and the force exerted by the fluid on the rigid body structure causes ...
+
+placeholder for a geometry overview
```
TODO

DavidSCN commented:
I added now a README and the tutorial is ready to merge from my side. 

> Or maybe reducing the (output) time step. We mainly want to change the visual effect here.

I like the visual effect :D

MakisH commented:
> I added now a README and the tutorial is ready to merge from my side.
> 
> > Or maybe reducing the (output) time step. We mainly want to change the visual effect here.
> 
> I like the visual effect :D

For me it feels that the flap moves too sharply with every plotting step. Usually we have more smooth transitions. But this can indeed be a matter of preference.

uekerman commented:
I also think we need a smaller timestep size:
* The displacement jumps are too big and, thus, accuracy is probably weak. Might be a strange "first impression".
* The visual effect.  
* Quasi-Newton coupling needs more similarity of the coupled system between the current and the last timestep.

Furthermore, the fluid side is rather unstable. Many inner iterations are needed. Small changes in the parameters can let the case crash (e.g. smaller dt + parallel coupling). We need a setup that is rather insensitive to changes, robustness-wise. I guess users would want to play around with such things.

What worked well for me was sth like:
* density fluid 1000
* nu 0.001
* density solid 10000
* dt = 0.005
* T= 2.5
... but then the displacement gets too big. I guess we will need some sort of rotational spring that works against the fluid force.

DavidSCN commented:
I added now an implicit rotational spring. After (too) many test runs, I conclude the following: The spring doesn't change the general problem. We have either stable setups or unstable setups. I guess there is even an analytic solution for some characteristic numbers which could tell us the stability region. If we want to see a visual effect in the results, the setups are (after a certain amount of time) unstable, for stable setups, the effect is 'only' visible in the watchpoint. Therefore, I added now a 'switch' in the middle of the simulation, which stiffens the spring so that the unstable initial setup is stabilized. The advantage would be, that you can really play around with the parameters and the simulation is rather stable. If it is not stable, the reason should be clear enough for the user. 
As always, feel free to try it yourself. I just realized that I need to update the given formulas.

uekerman commented:
I think the stiffness switch is a good solution. With all that, however, the simulation became a little bit too slow maybe. I'll try more things.

uekerman commented:
With a few tweaks, we are now done to 30s compute time for adapter + preCICE in debug mode. :tada:

DavidSCN commented:
> With a few tweaks, we are now done to 30s compute time for adapter + preCICE in debug mode. :tada:

Sounds good. I still had some points on my list for the README, in particular the formulas and an explanation for the rigid body model as well as the the explanation of the additional spring. Should this go in a separate PR or are you fine with the current state?

uekerman commented:
Feel free to open another PR or directly push to develop if the changes are pure extensions.
